lib/IDE/CodeCompletion.cpp

```diff
@@ -890,6 +890,25 @@ static void addObserverKeywords(CodeCompletionResultSink &Sink) {
   addKeyword(Sink, "didSet", CodeCompletionKeywordKind::None);
 }
 
+static void addKeywordsAfterReturn(CodeCompletionResultSink &Sink, DeclContext *DC) {
+  // `return nil` is not actually represented as a `ReturnExpr` in the AST but
+  // gets translated to a `FailStmt`. We thus can't produce the 'nil' completion
+  // using the solver-based implementation. Add the result manually.
+  if (auto ctor = dyn_cast_or_null<ConstructorDecl>(DC->getAsDecl())) {
```

Comment:
I don't really understand this statement. Since we are completing `nil`, the I think AST doesn't have it yet. So the AST should be like `(return_stmt (code_completion_expr))`. As far as I can see from the code, only `return nil` is converted to `FailExpr` but non-`nil` isn't.
As long as `CompletionLookup::addValueLiteralCompletions()` is called, `nil` seems to be added unconditionally. Could you explain why it's not the case?

Reply:
Oh, we don’t even get to type checking (and thus `sawSolutionImpl`, which would trigger `CompletionLookup`) because `PreCheckReturnStmtRequest` fails. I thought about changing `PreCheckStmtRequest` to allow `return #^COMPLETE^#` but I thought that allowing `return` statements inside initializers might cause issues further down the line if initializer type checking assumes that all `return`s have been translated to `FailStmt`.

https://github.com/apple/swift/blob/3687ddfac65a16c0cec61d7ab54a2c7c1760c330/lib/Sema/TypeCheckStmt.cpp#L1709-L1715